src/test-provider/test-manager.ts

```diff
@@ -0,0 +1,327 @@
+  TestDiscoveryResult,
+  TestItemMetadata,
+  TargetName,
+  TargetUri,
+} from "./types";
+
+export function createTestManager(
+  client: LanguageClient,
+  isDisabled: boolean
+): TestManager {
+  return new TestManager(client, isDisabled);
+}
+
+/**
+ * Get all descendant suites of a given TestItem.
+ * For build target it returns all test suite within it, same for package.
+ * If is called with TestItem which is a suite, then returns itself
+ */
+export function getTestItemChildren(test: vscode.TestItem): vscode.TestItem[] {
+  let children: vscode.TestItem[] = [];
+  if (testMetadata.get(test)?.kind === "suite") {
+    return [test];
+  } else {
+    test.children.forEach((child) => {
+      const descendants = getTestItemChildren(child);
+      children = [...children, ...descendants];
+    });
+    return children;
+  }
+}
+
+// this id is used to mark DAP sessions created by TestController
+// thanks to that debug tracker knows which requests it should track and gather results
+const testRunnerId = "scala-dap-test-runner";
+// mapping between TestItem and its metadata
+const testMetadata = new WeakMap<vscode.TestItem, TestItemMetadata>();
+const testSuiteResultMap = new Map<string, TestSuiteResult[]>();
+
+const getTestMetadata = (testItem: vscode.TestItem) =>
+  testMetadata.get(testItem);
+
+/**
+ * Register tracker which tracks all DAP sessions which are started with @constant {testRunnerId} kind.
+ * Dap sends execution result for every suite included in TestRun in a special event of kind 'testResult'.
+ * Tracker has to capture these events and store them all in map under debug session id as a key.
+ */
+vscode.debug.registerDebugAdapterTrackerFactory("scala", {
+  createDebugAdapterTracker(session) {
+    if (session.configuration.kind === testRunnerId) {
+      return {
+        onWillStartSession: () => testSuiteResultMap.set(session.id, []),
+        onDidSendMessage: (msg: DapEvent) => {
+          if (
+            msg.event === "testResult" &&
+            msg.body.category === "testResult"
+          ) {
+            testSuiteResultMap.get(session.id)?.push(msg.body.data);
+          }
+        },
+      };
+    }
+  },
+});
+
+/**
+ * runHandler is a function which is called to start a TestRun. Depending on the run profile it may be ordinary run or a debug TestRun.
+ * It creates run queue which contains test supposed to be run and then for each entry it creates & run a debug session
+ */
+async function runHandler(
+  testController: TestController,
+  noDebug: boolean,
+  request: TestRunRequest,
+  token: CancellationToken
+): Promise<void> {
+  const run = testController.createTestRun(request);
+  const queue = createRunQueue(request);
+
+  for (const { test, data } of queue) {
+    if (token.isCancellationRequested) {
+      run.skipped(test);
+    } else {
+      run.started(test);
+      const children = getTestItemChildren(test);
+      children.forEach((c) => run.enqueued(c));
+      await commands
```

Comment:
I would prefer not mixing async/await and the `then` api of promises if possible.
Also in general this expression is very long and I think it would benefit in readability if broken down in multiple smaller expressions

Reply:
Ok, so followng is a good way to go?
```typescript
try {
  const debugSession = await commands...
  const wasStarted = await ...
  await analyzeResults()
} catch(error) {
}
```